src/OrleansProviders/Storage/MemoryStorage.cs

```diff
@@ -33,67 +36,47 @@ namespace Orleans.Storage
     /// </code>
     /// </example>
     [DebuggerDisplay("MemoryStore:{Name}")]
-    public class MemoryStorage : IStorageProvider
+    public class MemoryGrainStorage : IGrainStorage, ILifecycleParticipant<ISiloLifecycle>
     {
-        /// <summary>
-        /// Default number of queue storage grains.
-        /// </summary>
-        public const int NumStorageGrainsDefaultValue = 10;
-        /// <summary>
-        /// Config string name for number of queue storage grains.
-        /// </summary>
-        public const string NumStorageGrainsPropertyName = "NumStorageGrains";
-        private int numStorageGrains;
+        private MemoryGrainStorageOptions options;
         private const string STATE_STORE_NAME = "MemoryStorage";
         private Lazy<IMemoryStorageGrain>[] storageGrains;
         private ILogger logger;
+        private IGrainFactory grainFactory;
+
         /// <summary> Name of this storage provider instance. </summary>
-        /// <see cref="IProvider.Name"/>
-        public string Name { get; private set; }
+        private readonly string name;
 
         /// <summary> Default constructor. </summary>
-        public MemoryStorage()
-            : this(NumStorageGrainsDefaultValue)
+        public MemoryGrainStorage(string name, MemoryGrainStorageOptions options, ILoggerFactory loggerFactory, IGrainFactory grainFactory)
         {
+            this.options = options;
+            this.name = name;
+            this.logger = loggerFactory.CreateLogger($"{this.GetType().FullName}.{name}");
+            this.grainFactory = grainFactory;
         }
 
-        /// <summary> Constructor - use the specificed number of store grains. </summary>
-        /// <param name="numStoreGrains">Number of store grains to use.</param>
-        protected MemoryStorage(int numStoreGrains)
-        {
-            numStorageGrains = numStoreGrains;
-        }
 
         #region IStorageProvider methods
 
         /// <summary> Initialization function for this storage provider. </summary>
-        /// <see cref="IProvider.Init"/>
-        public virtual Task Init(string name, IProviderRuntime providerRuntime, IProviderConfiguration config)
+        protected virtual Task Init(CancellationToken tk)
```

Comment:
None of this initialization is asynchronous.  Should probably all be done in constructor and not be a lifecycle participant at all.